raiden/transfer/architecture.py

```diff
@@ -272,6 +295,66 @@ def dispatch(self, state_change: StateChange) -> List[Event]:
 
         return events
 
+    def detect_balance_proof_change(self, old_state, current_state):
+        """Compare two states for any changed received balance_proofs"""
+        if old_state == current_state:
+            return
+        # Assuming, there is no transition None->new balance proof:
+        if not hasattr(old_state, 'identifiers_to_paymentnetworks'):
+            return
+        for old_payment_network_identifier, current_payment_network_identifier in zip(
```

Comment:
I don't think this works as you expect, `identifiers_to_paymentnetworks` are dictionaries, and iterating over them returns the dict keys in order of insertion (python 3.6+):

```python
>>> d = {1:1, 2:2}
>>> d2 = {2:2, 1:1}
>>> list(zip(d, d2))
[(1, 2), (2, 1)]
```

Which for all purposes is undefined at this point.

Reply:
Since this is getting the *new* balance proofs, it just needs to get the keys from `current_state.identifiers_to_paymentnetworks`, if there is no corresponding key in `old_state.identifiers_to_paymentnetworks` then we know all BPs are new.